### ISSUE TITLE
REOPEN: Issue with "key" statement missing for read-only "list"

### ISSUE THREAD

**Issue by GermanoSMO** [ict-strauss/COP#6]
By mistake I closed the issue #5 which actually is not yet addressed.

I resume here the point:

The swagger plugin requires mandatorily a "key" statement for a "list" in ALL conditions, including the case when the list itself is not configurable as a whole  (i.e. all of its elements can not be configured). This is not compliant with RFC6020 (and also with the upcoming YANG 1.1 described in draft RFC6020bis).

According to the RFC6020 definition (not changed in RFC6020bis), we have:

> 7.8.2.  The list's key Statement
> 
>  The "key" statement, (which) **MUST** be present if the list represents
>   configuration, and **MAY** be present otherwise

It's explicitly state then that the "key" statement is considered as _optional_ in case a list is not representing a configuration, i.e. in case a list is characterized by the "config false;" statement at the parent level for all of its leaves or it is a descendant of (contained by) a not-configurable node.

pyang 1.5 behaves consistently with this definition. I tried the following snippets and pyang in fact does not return any error for all of them. For your convenience, I have resumed the high level condition for the YANG code to be valid.

Condition A):
"List" not configurable due to inheritance/containment, i.e. the list is a descendant of a node characterized by "config false". Notice that it's not necessary that the list is a direct child of the non-configurable node: it may be at any level of containment with respect to the non-configurable ancestor.

``` yang
// SNIPPET No. 1
// "list" as direct child of a non-configurable node
container b {
    config false;     // container b is not configurable
    list beta {     
        leaf betaId {
            type uint8;
        }
    leaf foo {
        type int64;
           }
         }
}

// SNIPPET No. 2
// "list" as generic descendant of a non-configurable node (inherits the non configurability)
container a {
    config false;          // container a and all of its descendants are not configurable
    container b {
        list beta {     // list beta is not a direct child of "a", but is a descendant
            leaf betaId {
                type uint8;
            }
            leaf foo {
                type int64;
            }
        }
    }
}       
```

Condition B):
The "list" node is not-configurable as a whole (the parent/ancestor containers may be instead configurable)

``` yang
// SNIPPET No. 3
// A not-configurable "list" as descendant of a configurable node
    container a {       // the container a includes configurable nodes
        leaf dummy {
            type string;
            }
        list beta {
            config false;        // the list beta as a whole is not configurable
            leaf betaId {
                type uint8;
            }
            leaf foo {
                type int64;
            }
        }
    }       
```

IMPORTANT:
The "key" statement instead is mandatory in this case where although all the elements of the list are not-configurable, the list itself is _potentially_ configurable. For the following snippet, in fact, pyang returns "error: the list needs at least one key"

``` yang
// SNIPPET No. 4
// pyang complains about the missing "key" statement for the list
    container a {          // the container a includes configurable nodes
        leaf dummy {
            type string;
            }
        list beta {        // there's not a "config false" statement for the list beta
            leaf betaId {
                config false;
                type uint8;
            }
            leaf foo {
                config false;
                type int64;
            }
        }
    }       
```

So, you have two precise conditions to match (list as descendant of a node with "config false"; list with "config false") to align the swagger plugin behaviour to the valid YANG code, consistently with the results of pyang parsing.

**Comment by arthurMll** [comment]
Hi,
sorry for the delay on answering to your issue. I committed some changes in the swagger pyang-plugin in order to align it to the RFC. I think they correct the swagger generation models, at least I tried it with the snippets you post and it works great. 

Thanks you for your comment, this one in particular for the level of detail and the provided examples. Feel free to continue helping us to improve the code.

Regards,
A.
